File: app/src/main/java/com/bopr/android/smailer/sync/SyncService.java
package com.bopr.android.smailer.sync;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;

import androidx.annotation.Nullable;

/**
 * Service that returns an IBinder for the sync adapter class, allowing the sync adapter
 * framework to call onPerformSync()
 * <p>
 * Required by synchronization framework.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */

/* To debug it (to put breakpoints) remove android:process=":sync" from AndroidManifest */

public class SyncService extends Service {

    private static final Object lock = new Object();
    private static SyncAdapter adapter;

    @Override
    public void onCreate() {
        synchronized (lock) {
            if (adapter == null) {
                adapter = new SyncAdapter(getApplicationContext(), true);
            }
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return adapter.getSyncAdapterBinder();
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/sync/SyncService.kt
package com.bopr.android.smailer.sync

import android.app.IntentService
import android.content.Context
import android.content.Intent
import com.bopr.android.smailer.Database
import com.bopr.android.smailer.Settings
import com.bopr.android.smailer.Settings.Companion.PREF_SENDER_ACCOUNT
import com.bopr.android.smailer.Settings.Companion.PREF_SYNC_ENABLED
import com.bopr.android.smailer.util.getAccount
import org.slf4j.LoggerFactory

class SyncService : IntentService("google-drive-sync") {

    override fun onHandleIntent(intent: Intent?) {
        try {
            val settings = Settings(this)
            if (settings.getBoolean(PREF_SYNC_ENABLED)) {
                val account = getAccount(settings.getString(PREF_SENDER_ACCOUNT))
                account?.let {
                    Database(this).use {
                        Synchronizer(this, account, it).sync()
                    }
                } ?: log.warn("No sync account")
            }

            log.debug("Synchronized")
        } catch (x: Exception) {
            log.warn("Synchronization failed ", x)
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger("SyncService")

        fun startSyncService(context: Context) {
            log.debug("Starting service")

            context.startService(Intent(context, SyncService::class.java))
        }
    }
}